### PR TITLE
Handling primitive projections in canonical structures.

### PR DESCRIPTION
**Pull request by ppedrot**
This patch should allow to use primitive projections with canonical structures, as requested by @robbertkrebbers on Coq-club. For now there is no test included, because the bug report has not yet been submitted to Bugzilla, the latter having been down for a few days.

I think it deserves a backport to 8.7 as it is clearly a bugfix.

### CODE REVIEW

File: pretyping/evarconv.ml
Comment (Zimmi48) [inline pretyping/evarconv.ml]: Isn't it the kind of pattern that we are supposed to get rid of in the code? So as not to forget, let's say, to adapt the code of canonical structures whenever primitive projections are added?
BTW, isn't there an OCaml warning which could be activated to prevent this kind of style?

Comment (maximedenes) [inline pretyping/evarconv.ml]: I had a PR that was trying to remove those patterns on terms, but I'll need to rebase it.

Comment (SkySkimmer) [inline pretyping/evarconv.ml]: Warning 4 does that. It is disabled because it is triggered by 2855 lines (https://gist.github.com/SkySkimmer/4591058185a10c6f09016afd66d978d1).
That isn't to say it couldn't be turned on, I was trying to be conservative in #582.

Comment (Zimmi48) [inline pretyping/evarconv.ml]: Cool. Well, @maximedenes whenever you finish rebasing this PR, it would be great if you could turn warning 4 on.

Comment (maximedenes) [inline pretyping/evarconv.ml]: The PR removes such patterns only on terms, so I don't think it makes it possible to turn warning 4 on in practice.

Comment (Zimmi48) [inline pretyping/evarconv.ml]: Too bad. Well, we'll have to see how many are left after these, and where these are located, and if it is reasonable to e.g. activate warning 4 on some files / folders only.

**Comment by robbertkrebbers** [comment]
Thanks for the pull request!

Unfortunately, there seem to be more issues with canonical structures and primitive projections. See below for a piece of code that still fails with this patch:

```coq
Set Primitive Projections.

Structure semigroup := SemiGroup {
  sg_car :> Type;
  sg_op : sg_car -> sg_car -> sg_car;
}.

Structure monoid := Monoid {
  monoid_car :> Type;
  monoid_op : monoid_car -> monoid_car -> monoid_car;
  monoid_unit : monoid_car;
  monoid_left_id x : monoid_op monoid_unit x = x;
}.

Coercion monoid_sg (X : monoid) : semigroup :=
  SemiGroup (monoid_car X) (monoid_op X).
Canonical Structure monoid_sg.

Canonical Structure nat_sg := SemiGroup nat plus.
Canonical Structure nat_monoid := Monoid nat plus 0 plus_O_n.

Lemma foo (x : nat) : 0 + x = x.
Proof. Fail apply monoid_left_id.

(*
The command has indeed failed with message:
In environment
x : nat
Unable to unify
 "@eq (monoid_car ?M158) (monoid_op ?M158 (monoid_unit ?M158) ?M159) ?M159"
with "@eq nat (Nat.add O x) x".
*)
```

**Comment by ejgallego** [comment]
Again you may want to use `apply: monoid_left_id` or an equivalent.

**Comment by ejgallego** [comment]
To add a bit more context to my comment, IIUC `apply` uses some "variable technology" that is going to be removed in the near future, (c.f. #539) thus, fixing issues with `apply` may not be of very high priority until such changes land in `master`.

**Comment by robbertkrebbers** [comment]
> Again you may want to use apply: monoid_left_id or an equivalent.

That does not work in my actual code, where the problem occurs during type class inference, and there is no way of instructing type class inference to use `apply:`.

**Comment by ppedrot** [comment]
I just realized that the patch was not properly handling the case where the record type had parameters. This has been fixed by usual type-based expansion of the argument. I'm not a canonical structure user, so I don't really know how to design a canonical test that is supposed to work for structures with parameters.

**Comment by mattam82** [comment]
Indeed, that's the simplest way to fix it, using the compatibility layer. Hopefully that's enough but we should really have test-cases to check

**Comment by robbertkrebbers** [comment]
FWIW: I created a bug report, see https://coq.inria.fr/bugs/show_bug.cgi?id=5692

It contains some test cases too.

**Comment by ppedrot** [comment]
@mattam82 I'm wondering how much the expansion process is complete in the context of unification. Indeed, it fails if it cannot give the corresponding inductive type to the argument of the projection, but that's somewhat brutal. We could generate evars for parameters on the fly and postpone their unification. I don't know if this could have anything to do with the failing example of @robbertkrebbers...

**Comment by mattam82** [comment]
I've added more fixes here: https://github.com/mattam82/coq/commit/7a86afaa00ab78600264462d013ce3f973affb66 

The problem was that Proj(p,c) was not recognized as an open canonical projection in unification.ml, while the rest of the code works assuming that (evarconv and unification share this code actually, only this open canonical projection test is not done in evarconv).

I'm not sure I understand your worry @ppedrot : if Proj(p,c) is well "pretyped" (going through pretyping.ml), then it's an invariant that "c" has an inductive type.

**Review by mattam82 — APPROVED** [review]
With this set of patches, I think it solves the problem of unsupported primitive projections in canonical structures w.r.t. unification.ml

**Comment by ppedrot** [comment]
> I'm not sure I understand your worry @ppedrot : if Proj(p,c) is well "pretyped" (going through pretyping.ml), then it's an invariant that "c" has an inductive type.

Can't the type of `c` also be an evar, or any type blocked by an unresolved evar which would reduce to the corresponding inductive type once the evars are solved?

**Comment by ppedrot** [comment]
I've added the relevant commit by @mattam82, and the tests provided by @robbertkrebbers. This PR should be ready to merge and backport once the usual tests have passed.

**Comment by robbertkrebbers** [comment]
Thanks a lot for these patches, when I tried enabling primitive projections for selected parts of Iris, lots more goes through, but some issues still remain. I will try to create more minimized bug reports these days.

**Comment by robbertkrebbers** [comment]
See below for a minimized example in which primitive projections and canonical
structures still do not interact appropriately. Note that this example makes use
of the ssreflect rewrite tactic, but since ssreflect will be part of Coq 8.7, I
guess I should report this problem here, and not at the ssreflect bug tracker.

```coq
From Coq Require Import ZArith.
From Coq.ssr Require Import ssreflect.
Set Primitive Projections.

Structure semigroup := SemiGroup {
  sg_car :> Type;
  sg_op : sg_car -> sg_car -> sg_car;
}.

Structure group := Something {
  group_car :> Type;
  group_op : group_car -> group_car -> group_car;
  group_neg : group_car -> group_car;
  group_neg_op' x y : group_neg (group_op x y) = group_op (group_neg x) (group_neg y)
}.

Coercion group_sg (X : group) : semigroup :=
  SemiGroup (group_car X) (group_op X).
Canonical Structure group_sg.

Lemma group_neg_op (X : group) (x y : X) :
  group_neg _ (sg_op _ x y) = sg_op _ (group_neg _ x) (group_neg _ y).
Proof. apply group_neg_op'. Qed.

Canonical Structure Z_sg := SemiGroup Z Z.add .
Canonical Structure Z_group := Something Z Z.add Z.opp Z.opp_add_distr.

Lemma foo (x y z : Z) :
  sg_op Z_sg (group_neg Z_group x) (group_neg Z_group y) = z.
Proof.
  rewrite -group_neg_op.
(*
Ltac call to "rewrite (ssrrwargs) (ssrclauses)" failed.
Ltac call to "rewrite (ssrrwargs) (ssrclauses)" failed.
The RHS of group_neg_op

(sg_op (group_sg _) (group_neg _ _) (group_neg _ _))
does not match any subterm of the goal
*)
```

**Comment by robbertkrebbers** [comment]
FWIW: see bugzilla [#5697](https://coq.inria.fr/bugs/show_bug.cgi?id=5697), [#5698](https://coq.inria.fr/bugs/show_bug.cgi?id=5698) and [#5699](https://coq.inria.fr/bugs/show_bug.cgi?id=5699) for some other primitive projection related bugs that I discovered in Iris.

**Comment by robbertkrebbers** [comment]
Ping @gares since the above possibly contains a bug in the ssreflect plugin.

**Comment by gares** [comment]
@robbertkrebbers indeed... I've made a PR on top of PMP's one
  https://github.com/ppedrot/coq/pull/3
it makes the test go trough.

**Comment by robbertkrebbers** [comment]
Thanks, @gares!

With the fixes so far, I am able to enable primitive projections for crucial some parts of Iris, and get a significant performance improvement (up to 25%) on some files. Chapeau!

(I have to work around Coq bugs #5699 and #5706, but that is pretty minor).

Still, trying to enable it for the whole of the Iris development results in some problems. I will continue to report minimized bugs these days.

**Comment by ppedrot** [comment]
Shouldn't we merge @gares patch as a separate PR? It is a bugfix for another component, isn't it?

**Comment by robbertkrebbers** [comment]
> Shouldn't we merge @gares patch as a separate PR? It is a bugfix for another component, isn't it?

If needed, I can also make a separate bug report for it.

**Comment by gares** [comment]
I can make a separate PR with 1 commit, but what would be the gain?
 @maximedenes what do you suggest?

**Comment by robbertkrebbers** [comment]
> Still, trying to enable it for the whole of the Iris development results in some problems. I will continue to report minimized bugs these days.

After some investigation, it seems the remaining problems (see bugzilla #5707 and #5708, but probably more to come) are unrelated to canonical structures. That said, I would be very happy if this PR + @gares patch could be merged.

**Comment by ppedrot** [comment]
Travis fails with this PR on HoTT. Is it because of the patch or is there something wrong with HoTT currently?

**Comment by Zimmi48** [comment]
No HoTT is fine currently.

**Comment by ppedrot** [comment]
Indeed, this patch breaks HoTT. The following is enough to fix it though:
```
diff --git a/theories/Categories/ChainCategory.v b/theories/Categories/ChainCategory.v
index ee2abff8a..8a6967113 100644
--- a/theories/Categories/ChainCategory.v
+++ b/theories/Categories/ChainCategory.v
@@ -81,7 +81,8 @@ Module Export Univalent.
   Proof.
     intros s d.
     refine (isequiv_iff_hprop _ _).
-    { refine (trunc_equiv' _ (issig_isomorphic _ _ _)). }
+    { refine (trunc_equiv' _ (issig_isomorphic _ _ _)).
+      simpl; refine _. }
     { intro m; apply leq_antisym; apply m. }
   Defined.
```
It looks like some reduction on primitive projections was not perform, leading to a failure in class resolution. What do you think, @mattam82?

**Comment by mattam82** [comment]
Don't really have time right now, but if you show the goal before simpl I might have an idea!

**Comment by robbertkrebbers** [comment]
The goal before `simpl`:

```coq
s, d : omega
______________________________________(1/1)
IsHProp {m : morphism omega s d & IsIsomorphism m}
```

When trying to use `apply _` on that goal:

```coq
Cannot infer this placeholder of type "IsHProp {m : morphism omega s d & IsIsomorphism m}" (no type class instance
found) in environment:
s, d : omega
```

The goal after `simpl`:

```coq
s, d : omega
______________________________________(1/1)
IsHProp {m : s <= d & IsIsomorphism m}
```

For completeness sake, also before `simpl` with `Printing All`:

```coq
IsTrunc (trunc_S minus_two) (@sig (morphism omega s d) (fun m : morphism omega s d => @IsIsomorphism omega s d m))
```

And after `simpl`:

```coq
IsTrunc (trunc_S minus_two)
  (@sig (@trunctype_type _ (dhprop_hprop (leq s d)))
     (fun m : @trunctype_type _ (dhprop_hprop (leq s d)) => @IsIsomorphism omega s d m))
```

And moreover, `morphism` is a projection of a record with primitive projections:

```coq
NonCumulative Record PreCategory : Type := Build_PreCategory'
  { object : Type;
    morphism : object -> object -> Type;
    identity : forall x : object, morphism x x;
    compose : forall s d d' : object, morphism d d' -> morphism s d -> morphism s d';
    associativity : forall (x1 x2 x3 x4 : object) (m1 : morphism x1 x2) (m2 : morphism x2 x3) (m3 : morphism x3 x4),
                    compose x1 x2 x4 (compose x2 x3 x4 m3 m2) m1 = compose x1 x3 x4 m3 (compose x1 x2 x3 m2 m1);
    associativity_sym : forall (x1 x2 x3 x4 : object) (m1 : morphism x1 x2) (m2 : morphism x2 x3) 
                        (m3 : morphism x3 x4),
                        compose x1 x3 x4 m3 (compose x1 x2 x3 m2 m1) = compose x1 x2 x4 (compose x2 x3 x4 m3 m2) m1;
    left_identity : forall (a b : object) (f : morphism a b), compose a b b (identity b) f = f;
    right_identity : forall (a b : object) (f : morphism a b), compose a a b f (identity a) = f;
    identity_identity : forall x : object, compose x x x (identity x) (identity x) = identity x;
    trunc_morphism : forall s d : object, IsHSet (morphism s d) }
```

**Comment by robbertkrebbers** [comment]
When comparing the type class output before `simpl` with that of after `simpl`, the problem seems to be here:

```coq
Goal forall s d : omega, IsHProp (morphism omega s d).
intros.
simpl. simple apply istrunc_trunctype_type. (* Works *)

Restart.
intros.
Fail simple apply istrunc_trunctype_type. (* Unable to find an instance for the variable _. *)

(* The goal with Printing all:

  IsTrunc (trunc_S minus_two) (morphism omega s d)

*)
Check istrunc_trunctype_type.
(*

  istrunc_trunctype_type
     : forall (n : trunc_index) (t : TruncType n), IsTrunc n (@trunctype_type _ t)
*)
```

**Comment by ppedrot** [comment]
@mattam82 Ping. It's probably important to backport this patch in 8.7, so the sooner the breakage is assessed, the better.

**Comment by RalfJung** [comment]
This patch makes it so that we can finally use primitive projections in Iris, fixing some severe performance problems. It would be rather sad to have to wait for another 6 months until we can actually use primitive projections. Any chance this can go into 8.7?

**Comment by Zimmi48** [comment]
@mattam82 This is apparently dependent on you.

**Comment by mattam82** [comment]
Ok, I can't promise to analyze the latest bug before Thursday but I'll try to have a look.

**Comment by maximedenes** [comment]
> Ok, I can't promise to analyze the latest bug before Thursday but I'll try to have a look.

Sorry to bring sad news, but if not ready today, it won't make it to 8.7 final (like other changes).

**Comment by RalfJung** [comment]
Given that the fix for HoTT is a one-line diff, would it be a problem to accept the PR in its current state?

**Comment by ppedrot** [comment]
Do we have other users of primitive projections? If so, we need to test on their developments as well.

**Comment by RalfJung** [comment]
Iris uses primitive projections for sealing: <https://gitlab.mpi-sws.org/robbertkrebbers/coq-stdpp/blob/master/theories/base.v#L20>

Using primitive projections for anything else (like our algebraic classes) does not work with Coq 8.6 for us.

**Comment by mattam82** [comment]
That's not ideal, it's a regression right? I'm looking into it and it seems to be due to the fact that [morphism] is a folded primitive projection here, once you [unfold] it [refine _] works.

**Comment by RalfJung** [comment]
Actually it turns out we could use primitive projections in Iris for two more records currently, though it hardly makes any difference for compile times: <https://gitlab.mpi-sws.org/FP/iris-coq/merge_requests/69>.

We could test this PR against that branch as well. However, that may not actually end up using primitive projections in any interesting way.

**Comment by mattam82** [comment]
I elapsed my time for looking at this for now, but it's definitely a
problem of folded/unfolded versions of primitive projections.

**Comment by mattam82** [comment]
Ok, the fix to HoTT is right: CS resolution works "too well" now. What happens is that we get to the following unification problem:
```
trunctype_type ?M20796 ~= morphism omega s
```
Unification decides that it can try CS resolution, and finds a solution using the default canonical structure declared for TruncType. Except this default structure actually is kind of a hack, as it leaves a new IsTrunc _ typeclass instance as a side effect, which is not resolvable (but it's too late in w_merge or after to backtrack on this CS choice). What happened before the patch is that CS resolution was not launched or failing, and we instead unfolded morphism omega which reduces to trunctype_type ... so just FO unification succeeded. I believe one should just fix the script using unfold morphism; refine _. This kind of interaction of CS and typeclass resolution is quite fragile I think.

**Comment by ppedrot** [comment]
@mattam82 So what? Is this patch supposed to land in 8.7.1? Or are we postponing it to 8.8 because it breaks developments?

**Comment by mattam82** [comment]
It's a bugfix so I would consider it for 8.7.1

**Comment by maximedenes** [comment]
> It's a bugfix so I would consider it for 8.7.1

Minor releases are supposed to be 100% compatible, except for scripts relying on bugs. Was HoTT relying on a bug (as in violation of a documented spec)?

**Comment by mattam82** [comment]
Yes, it's because CS resolution didn't work with primitive projections that their script worked.

**Comment by mattam82** [comment]
I don't think we promise anything about the interaction of CS and typeclass resolution though, so as far as documention / spec goes, I would argue it falls into "primitive projections should work just like regular projections". The script would fail if regular projections were used instead.

**Comment by maximedenes** [comment]
> "primitive projections should work just like regular projections". The script would fail if regular projections were used instead.

Indeed, I find this argument easier to buy than the fact that the interaction between CS and typeclass resolution was not the "expected" one.

**Comment by ppedrot** [comment]
I think this is ready for merging.

**Comment by Zimmi48** [comment]
> I think this is ready for merging.

Did you do something for HoTT comaptibility?

**Comment by ppedrot** [comment]
I have the patch lying around, but I did not submit a PR yet. What's the protocol for synchronization?

**Comment by maximedenes** [comment]
Is the fix backward compatible (i.e. still compiles HoTT fine with 8.7.0)?

**Comment by ppedrot** [comment]
@maximedenes It can be made so, at least. Just replace the dot in my patch with a semicolon.

**Comment by maximedenes** [comment]
I'm wrapping up the release, so if you could submit a PR on HoTT with a backward compatible fix, there would be no sync issues :)

**Comment by Zimmi48** [comment]
Note that we are still testing @ejgallego's branch of HoTT so the patch will need to be integrated there as well (or maybe only there, I don't know).

**Comment by ppedrot** [comment]
@Zimmi48 Isn't it a serious issue that we're only testing @ejgallego's branch? How come HoTT master hasn't kept up with Coq changes? Also, I guess a merge into our local branch should be enough, hopefully...

**Comment by Zimmi48** [comment]
> @Zimmi48 Isn't it a serious issue that we're only testing @ejgallego's branch? How come HoTT master hasn't kept up with Coq changes? Also, I guess a merge into our local branch should be enough, hopefully...

Sure. I don't know what was the initial problem that justified not tracking upstream directly but I agree it should be fixed ASAP if we want to avoid problems and over-complications in the future.

**Comment by SkySkimmer** [comment]
The init plugins changed in 8.7 and HoTT has a stripped down copy of the init files so it needs a version for 8.6 and one for 8.7.

**Comment by ejgallego** [comment]
Even if we are testing a private HoTT branch, there is good communication with the HoTT team and we will improve the situation soon.

**Comment by ppedrot** [comment]
@ejgallego Can you cherry pick HoTT/HoTT#898 on your branch in the meantime? The fix is OK, it passes their Travis even on 8.6. This way we can merge this PR.

**Comment by ejgallego** [comment]
@ppedrot done but somehow the github interface doesn't update.

**Comment by ppedrot** [comment]
The patch in HoTT has been merged.

**Comment by robbertkrebbers** [comment]
Thanks for merging, and looking forward to seeing it appear in 8.7.1 :).

**Pull request merged**
